awscli/topics/s3-config.rst

```diff
@@ -0,0 +1,135 @@
+-----------------------
+
+**Default** - ``10``
+
+The ``aws s3`` transfer commands are multithreaded.  At any given time,
+multiple requests to Amazon S3 are in flight.  For example, if you are
+uploading a directory via ``aws s3 cp localdir s3://bucket/ --recursive``, the
+AWS CLI could be uploading the local files ``localdir/file1``,
+``localdir/file2``, and ``localdir/file3`` in parallel.  The
+``max_concurrent_requests`` specifies the maximum number of transfer commands
+that are allowed at any given time.
+
+You may need to change this value for a few reasons:
+
+* Decreasing this value - On some environments, the default of 10 concurrent
+  requests can overwhelm a system.  This may cause connection timeouts or
+  slow the responsiveness of the system.  Lowering this value will make the
+  S3 transfer commands less resource intensive.  The tradeoff is that
+  S3 transfers may take longer to complete.
+* Increasing this value - In some scenarios, you may want the S3 transfers
+  to complete as quickly as possible, using as much network bandwidth
+  as necessary.  In this scenario, the default number of concurrent requests
+  may not be sufficient to utilize all the network bandwidth available.
+  Increasing this value may improve the time it takes to complete an
+  S3 transfer.
+
+
+max_queue_size
+--------------
+
+**Default** - ``1000``
+
+The AWS CLI internally uses a producer consumer model, where we queue up S3
+tasks that are then executed by consumers, which in this case utilize a bound
+thread pool, controlled by ``max_concurrent_requests``.  The enqueuing rate
+can be much faster than the rate at which consumers are executing tasks.
+To avoid unbounded growth, the task queue size is capped to a specific size.
+This configuration value changes the value of that maximum number.
+
+You generally will not need to change this value.  This value also
+corresponds to the number of tasks we are aware of that need to be
+executed.  This means that by default we can only see 1000 tasks ahead.
+Until the S3 command knows the total number of tasks executed, the
+progress line will show a total of ``...``.  Increasing this value
+means that we will be able to more quickly know the total number of
+tasks needed, assuming that the enqueuing rate is quicker than the
+rate of task consumption.  The tradeoff is that a larger max queue
+size will require more memory.
+
+
+multipart_threshold
+-------------------
+
+**Default** - ``8MB``
+
+When uploading, downloading, or copying a file, the S3 commands
+will switch to multipart operations if the file reaches a given
+size threshold.  The ``multipart_threshold`` controls this value.
+You can specify this value in one of two ways:
+
+* The file size in bytes.  For example, ``1048576``.
+* The file size with a size suffix.  You can use ``KB``, ``MB``, ``GB``,
+  ``TB``.  For example: ``10MB``, ``1GB``.  Note that S3 imposes
+  constraints on valid values that can be used for multipart
+  operations.
```

Comment:
Might want to mention that there is a general maximum of a non multipart upload size of 5 GB.